### PR TITLE
Google Authorization

### PR DESCRIPTION
**Pull request by LanchevychMaxym**

### CODE REVIEW

**Review by tetyanakutsko — APPROVED** [review]

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
+﻿namespace EPlast.BLL.Models
+{
+    public class GoogleApiTokenInfo
+    {
+        /// <summary>
+        /// The Issuer Identifier for the Issuer of the response. Always https://accounts.google.com or accounts.google.com for Google ID tokens.
+        /// </summary>
+        public string iss { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:8]:
```suggestion
        public string Iss { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// Access token hash. Provides validation that the access token is tied to the identity token. If the ID token is issued with an access token in the server flow, this is always
+        /// included. This can be used as an alternate mechanism to protect against cross-site request forgery attacks, but if you follow Step 1 and Step 3 it is not necessary to verify the 
+        /// access token.
+        /// </summary>
+        public string at_hash { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:15]:
```suggestion
        public string AtHash { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// Identifies the audience that this ID token is intended for. It must be one of the OAuth 2.0 client IDs of your application.
+        /// </summary>
+        public string aud { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:20]:
```suggestion
        public string Aud { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// An identifier for the user, unique among all Google accounts and never reused. A Google account can have multiple emails at different points in time, but the sub value is never
+        /// changed. Use sub within your application as the unique-identifier key for the user.
+        /// </summary>
+        public string sub { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:26]:
```suggestion
        public string Sub { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// True if the user's e-mail address has been verified; otherwise false.
+        /// </summary>
+        public string email_verified { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:31]:
```suggestion
        public string EmailVerified { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// The client_id of the authorized presenter. This claim is only needed when the party requesting the ID token is not the same as the audience of the ID token. This may be the
+        /// case at Google for hybrid apps where a web application and Android app have a different client_id but share the same project.
+        /// </summary>
+        public string azp { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:37]:
```suggestion
        public string Azp { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// The user's email address. This may not be unique and is not suitable for use as a primary key. Provided only if your scope included the string "email".
+        /// </summary>
+        public string email { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:42]:
```suggestion
        public string Email { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// The time the ID token was issued, represented in Unix time (integer seconds).
+        /// </summary>
+        public string iat { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:47]:
```suggestion
        public string Iat { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// The time the ID token expires, represented in Unix time (integer seconds).
+        /// </summary>
+        public string exp { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:52]:
```suggestion
        public string Exp { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// The user's full name, in a displayable form. Might be provided when:
+        /// The request scope included the string "profile"
+        /// The ID token is returned from a token refresh
+        /// When name claims are present, you can use them to update your app's user records. Note that this claim is never guaranteed to be present.
+        /// </summary>
+        public string name { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:60]:
```suggestion
        public string Name { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// The URL of the user's profile picture. Might be provided when:
+        /// The request scope included the string "profile"
+        /// The ID token is returned from a token refresh
+        /// When picture claims are present, you can use them to update your app's user records. Note that this claim is never guaranteed to be present.
+        /// </summary>
+        public string picture { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:68]:
```suggestion
        public string Picture { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+        public string given_name { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:70]:
```suggestion
        public string GivenName { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+        public string family_name { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:72]:
```suggestion
        public string FamilyName { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+        public string locale { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:74]:
```suggestion
        public string Locale { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+        public string alg { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:76]:
```suggestion
        public string Alg { get; set; }
```

File: EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+        public string kid { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Models/GoogleApiTokenInfo.cs:78]:
```suggestion
        public string Kid { get; set; }
```

File: EPlast/EPlast.BLL/Services/Auth/AuthService.cs [hunk 17]
```diff
@@ -241,6 +246,59 @@ public async Task SendEmailReseting(string confirmationLink, ForgotPasswordDto f
...
         }
 
+        ///<inheritdoc/>
+        public async Task<User> GetGoogleUserAsync(string providerToken)
+        {
+            string GoogleApiTokenInfoUrl = "https://www.googleapis.com/oauth2/v3/tokeninfo?id_token={0}";
+            var httpClient = new HttpClient();
+            var requestUri = new Uri(string.Format(GoogleApiTokenInfoUrl, providerToken));
+
+            HttpResponseMessage httpResponseMessage = httpClient.GetAsync(requestUri).Result;
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Services/Auth/AuthService.cs:256]:
```suggestion
            HttpResponseMessage httpResponseMessage = await httpClient.GetAsync(requestUri)
```
Why don't we use this?

File: EPlast/EPlast.BLL/Services/Auth/AuthService.cs [hunk 18]
```diff
@@ -241,6 +246,59 @@ public async Task SendEmailReseting(string confirmationLink, ForgotPasswordDto f
...
+
+            if (httpResponseMessage.StatusCode != HttpStatusCode.OK)
+            {
+                throw new Exception();
+            }
+
+            var response = httpResponseMessage.Content.ReadAsStringAsync().Result;
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Services/Auth/AuthService.cs:263]: ?

File: EPlast/EPlast.WebApi/Controllers/AuthController.cs [hunk 19]
```diff
@@ -11,6 +13,12 @@
 using Microsoft.Extensions.Localization;
 using System.Threading.Tasks;
 using System.Web;
+using EPlast.BLL.DTO.UserProfiles;
+using EPlast.DataAccess.Entities;
```
Comment (voroniak) [inline EPlast/EPlast.WebApi/Controllers/AuthController.cs:17]: Why do we have reference to Data Access?

**Review by voroniak — CHANGES_REQUESTED** [review]

File: EPlast/EPlast.WebApi/Controllers/AuthController.cs (hunk 19 above)
Comment (LanchevychMaxym) [reply EPlast/EPlast.WebApi/Controllers/AuthController.cs]: I've already fixed it, thanks.

**Review by MarianLabensky — APPROVED** [review]

**Review by tarasvolosheniuk — APPROVED** [review]

**Review by MykhailoKlapchuk — APPROVED** [review]

**Review by voroniak — APPROVED** [review]

**Pull request merged**